### DISCUSSION TITLE
Possible DoS by deposit whitelist users

### DISCUSSION THREAD

**Discussion by alexaunter** [galactic3/ft-lockup-contract#38] [Q&A] [answered]
deposit whitelist account can prevent anyone from claiming their money.

Steps:

1. owner.near creates unterminatable lockup for alice.near on 60000 TOKEN
2. owner decides to prevent user from receiving the money
3. owner.near create 300+ additional lockups for alice.near of minimal possible amount (0.0000000001 TOKEN) each
4. time passes, alice.near decides to withdraw money (according to the schedule)
5. alice.near calls claim method, claim methods iterates over ALL user lockups, runs out of gas while reading 300+ lockups (300 state reads)
6. alice.near is unable to receive money, while it should be able to do so

Cost of attack: 300 * (0.0004 NEAR + 0.0000000001 TOKEN)
Sounds like an exploit, since it breaks guarantee of receiving the money.
Possible solution: add optional lockup_ids parameter to the claim method, to guarantee O(1) gas usage.

**Comment by alexaunter** [comment, marked as answer]
@frol, what actions should we make to mitigate this scenario

**Reply by frol** [reply]
@alexaunter As you suggested during the call, let's add a new method that can take a list of pairs: lockup id + amount to withdraw, thus we can control the amount of gas we spend.

cc @evgenykuzyakov
